src/cmd_line/commandLine.ts

```diff
@@ -8,7 +8,75 @@ import * as parser from './parser';
 import * as util from '../util';
 import { VimError, ErrorCode } from '../error';
 
+class CommandLineHistory {
+  private _history: string[] = [];
+  private _is_loaded: boolean = false;
+  private _filePath: string = '';
+
+  public add(command: string | undefined): void {
+    if (!command || command.length === 0) {
+      return;
+    }
+
+    let index: number = this._history.indexOf(command);
+    if (index !== -1) {
+      this._history.splice(index, 1);
+    }
+    this._history.unshift(command);
+
+    if (this._history.length > configuration.history) {
+      this._history.pop();
+    }
+  }
+
+  public get(): string[] {
+    if (!this._is_loaded) {
+      this.load();
+      this._is_loaded = true;
+    }
+
+    if (this._history.length > configuration.history) {
+      // resize history because "vim.history" is updated.
+      this._history = this._history.slice(0, configuration.history);
+      this.save();
+    }
+
+    return this._history;
+  }
+
+  public setFilePath(filePath: string): void {
+    this._filePath = filePath;
+  }
+
+  private load(): void {
+    const fs = require('fs');
+
+    if (!fs.existsSync(this._filePath)) {
+      return;
+    }
+
+    try {
+      let data = fs.readFileSync(this._filePath, 'utf-8');
+      let parsedData = JSON.parse(data);
+      if (Array.isArray(parsedData)) {
+        this._history = parsedData;
+      } else {
+        console.log('CommandLine: Failed to load history.');
+      }
+    } catch (e) {
+      console.error(e);
+    }
+  }
+
+  public save(): void {
+    const fs = require('fs');
+    fs.writeFileSync(this._filePath, JSON.stringify(this._history), 'utf-8');
```

Comment:
Should this be asynchronous? I'd prefer to avoid synchronous file calls in general, but in this case, I don't see any advantage to do it synchronously.

Perhaps the other file calls should also be synchronous too, but I don't have a strong opinion on that.

Reply:
I have an external HDD that takes a while to spin up. It looks like right now history is being written to the extension folder, but if it was ever made per-project, I'm worried these sync calls could cause a ton of lag. No way around it for loading history, but at least for saving, can this be run async at the same time as the command (e.g. don't await it)?

Similarly for load, there's no way around a slow disk read, but will the sync call cause the editor to freeze if loading takes a long time?

Reply:
Well, if it's made asynchronous, it shouldn't lock up the editor. It won't be able to display until it's done loading, but I don't think that's avoidable. It should only need to load once per VSCode session though.

I'm leaning towards making all the file operations asynchronous.

Reply:
Async please. Especially considering you are reading it from memory on the `gets` so there's no reason to block on the file write.

Reply:
I made file operations asynchronous.
0a635c35a225c39ed1dc2a8b6d1e5e6010183368

Is it too overworked to save the command executed while loading the file?